static/app/utils/useNavigate.tsx

```diff
@@ -1,28 +1,43 @@
-import {useContext} from 'react';
+import {useCallback, useEffect, useRef} from 'react';
 
-import {RouteContext} from 'sentry/views/routeContext';
+import {useRouteContext} from 'sentry/utils/useRouteContext';
 
 type NavigateOptions = {
   replace?: boolean;
   state?: any;
 };
 
 export function useNavigate() {
-  const route = useContext(RouteContext);
-  if (route === null) {
-    throw new Error('useNavigate called outside of routes provider');
-  }
-  const navigator = route?.router;
+  const route = useRouteContext();
 
-  const navigate = (to: string | number, options: NavigateOptions = {}) => {
-    if (typeof to === 'number') {
-      navigator.go(to);
-      return;
-    }
-    (!!options.replace ? navigator.replace : navigator.push)({
-      pathname: to,
-      state: options.state,
-    });
-  };
+  const navigator = route.router;
+  const activeRef = useRef(false);
+  useEffect(() => {
+    activeRef.current = true;
+  });
+  const navigate = useCallback(
+    (to: string | number, options: NavigateOptions = {}) => {
+      if (!activeRef.current) {
+        throw new Error(
+          `You should call navigate() in a React.useEffect(), not when your component is first rendered.`
```

Comment:
What is the reason for this limitation? It doesnt look obvious to me why this would not be allowed? I'm afraid that someone may mistakenly do things like and not test the behavior which will result in prod crashes :/
```tsx
const component = () => {
const navigate = useNavigate()

if(someRedirectCondition) {
  navigate()
}

return ....
}
```
For the error message - you can call useEffect in a useCallback too right? Maybe something like "Calling navigate before a component has finished rendering is not allowed."

Reply:
@JonasBa I stole that warning from react-router v6 https://github.com/remix-run/react-router/blob/main/packages/react-router/index.tsx#L534-L538 just to keep our usage of this hook consistent with an eventual upgrade to v6.

The recommended way is:

```js
const component = () => {
const navigate = useNavigate()

useEffect(() => {
  if(someRedirectCondition) {
    navigate("/")
  }
}, [])

return ....
}
```